### ISSUE TITLE
[Documentation] ⚠️ Regarding events/logs, I may have pointed you in the wrong direction during our call in such you should emit events at will.

### ISSUE THREAD

**Issue by psavelis** [kanna-coin/onchain-smart-contracts#14]
_Originally posted by @hbarcelos in https://github.com/kanna-coin/treasury-staking-monorepo/issues/13#issuecomment-1261246081_

        ⚠️ Regarding events/logs, I may have pointed you in the wrong direction during our call in such you should emit events at will.

I am using [this](https://github.com/djrtwo/evm-opcode-gas-costs/blob/master/opcode-gas-costs_EIP-150_revision-1e18248_2017-04-12.csv#L60-L64) as the source for accounting. Not all opcode costs are updated, but the cost of logs didn't change recently.

Events have 2 components:
1. Topics: the event signature and up to 3 `indexed` parameters, which are hashed using `keccak256`, therefore they take 32 bytes each.
2. Data: all non-`indexed` parameters, which are [ABI encoded](https://docs.soliditylang.org/en/latest/abi-spec.html).

Taking this event as an example:

```solidity
    event TransactionFee(
        address from,
        address to,
        uint256 amount,
        uint256 indexed date, /* TODO: Why is this field indexed? */
        uint256 fee
    );
```

For that reason, we will need to use the `LOG2` opcode. From the gas cost table, we have:

```
375 + 8 * (number of bytes in log data) + 2 * 375
```

We have:
- 2 topics: 1 event signature + 1 indexed parameter.
- 4 non-indexed parameters. Using the ABI encoding we will end up with $4*32=128$ bytes.

Applying the values to the equation above:  
$375+(8\times128)+(2\times375)=2149$ gas

At today's rates, this is equivalent to $0.07$ USD, but the actual value depends both on gas price and ETH prices.

In comparison, to store the same data in contract storage, it would cost up to $5\times22100=110500$ gas, or $3.67$ USD. That is $52\times$ more expensive than using events.

TL;DR: while events are way cheaper than storage, the cost is low, but not negligible. There is no need to not use events when they are needed, but they should be used with caution.

_Originally posted by @hbarcelos in https://github.com/kanna-coin/treasury-staking-monorepo/issues/13#issuecomment-1261246081_
